tools/pd-simulator/simulator/task.go

```diff
@@ -49,380 +53,462 @@ const (
-				regionID: regionID,
-				size:     region.GetApproximateSize(),
-				keys:     region.GetApproximateKeys(),
-				epoch:    epoch,
-				peer:     changePeer.GetPeer(),
-				// This two variables are used to simulate sending and receiving snapshot processes.
-				sendingStat:   newSnapshotState(region.GetApproximateSize(), generate),
-				receivingStat: newSnapshotState(region.GetApproximateSize(), receive),
+	switch {
+	case resp.GetChangePeer() != nil:
+		op, desc = changePeerToOperator(region, resp.GetChangePeer())
+	case resp.GetChangePeerV2() != nil:
+		cps := resp.GetChangePeerV2().GetChanges()
+		if len(cps) == 0 {
+			// leave joint state
+			desc = fmt.Sprintf("leave joint state for region %d", regionID)
+			op = &changePeerV2Leave{}
+		} else if len(cps) == 1 {
+			// original ChangePeer
+			op, desc = changePeerToOperator(region, cps[0])
+		} else {
+			// enter joint state, it can only contain PromoteLearner and DemoteVoter.
+			subDesc := make([]string, 0, len(cps))
+			cp2 := &changePeerV2Enter{}
+			for _, cp := range cps {
+				peer := cp.GetPeer()
+				subOp, _ := changePeerToOperator(region, cp)
+				switch subOp.(type) {
+				case *promoteLearner:
+					subDesc = append(subDesc, fmt.Sprintf("promote peer %+v", peer))
+					cp2.promoteLearners = append(cp2.promoteLearners, peer)
+				case *demoteVoter:
+					subDesc = append(subDesc, fmt.Sprintf("demote peer %+v", peer))
+					cp2.demoteVoters = append(cp2.demoteVoters, peer)
+				default:
+					simutil.Logger.Error("cannot exec AddPeer or RemovePeer when using joint state")
+					return nil
+				}
 			}
+			desc = fmt.Sprintf("%s for region %d", strings.Join(subDesc, ", "), regionID)
+			op = cp2
 		}
-	} else if resp.GetTransferLeader() != nil {
-		changePeer := resp.GetTransferLeader().GetPeer()
-		fromPeer := region.GetLeader()
-		return &transferLeader{
-			regionID: regionID,
-			epoch:    epoch,
-			fromPeer: fromPeer,
-			peer:     changePeer,
+	case resp.GetTransferLeader() != nil:
+		fromPeerStoreID := region.GetLeader().GetStoreId()
+		// When this field is included, it means that TiKV needs to decide the optimal Leader by itself.
+		toPeers := resp.GetTransferLeader().GetPeers()
+		// When no Peers are included, use Peer to build Peers of length 1.
+		if len(toPeers) == 0 {
+			toPeers = []*metapb.Peer{resp.GetTransferLeader().GetPeer()}
 		}
-	} else if resp.GetMerge() != nil {
+		desc = fmt.Sprintf("transfer leader from store %d to store %d", fromPeerStoreID, toPeers[0].GetStoreId())
+		op = &transferLeader{
+			fromPeerStoreID: fromPeerStoreID,
+			toPeers:         toPeers,
+		}
+	case resp.GetMerge() != nil:
 		targetRegion := resp.GetMerge().GetTarget()
-		return &mergeRegion{
-			regionID:     regionID,
-			epoch:        epoch,
-			targetRegion: targetRegion,
+		desc = fmt.Sprintf("merge region %d into %d", regionID, targetRegion.GetId())
+		op = &mergeRegion{targetRegion: targetRegion}
+	case resp.GetSplitRegion() != nil:
+		// TODO: support split region
```

Comment:
when will we support it?

Reply:
We need to discuss how to perform for the three different splits.